### PR TITLE
Refactor parser architecture in order to follow the cake pattern

### PR DESCRIPTION
**Pull request by JonathanGeoffroy**
Although we have implemented the cake pattern, mocking is still impossible for unknown reasons.
Tests on parser remain more functional than unit.

### CODE REVIEW

File: src/main/scala/fr/inria/hopla/ast/ASTField.scala [hunk 1]
```diff
@@ -5,5 +5,5 @@ package fr.inria.hopla.ast
  * @author Jérémy Bossut, Jonathan Geoffroy
  */
 class ASTField(name: String, fieldType: String) {
-  def getName = name
+  val getName = name
```
Comment (rouvoy) [inline src/main/scala/fr/inria/hopla/ast/ASTField.scala]: Avoid such getters and add `val` to the parameters of the class you want to access from outside.

File: src/main/scala/fr/inria/hopla/ast/ASTFile.scala [hunk 2]
```diff
@@ -38,8 +38,8 @@ abstract case class ASTFile(name: String) {
    * @param astTrait the trait implemented by this ASTFile
    */
   def withTrait(astTrait: ASTTrait): Unit = {
-    traits = traits + astTrait
-    astTrait.subFiles = astTrait.subFiles + this
+      traits = traits + astTrait
```
Comment (rouvoy) [inline src/main/scala/fr/inria/hopla/ast/ASTFile.scala]: `traits += astTrait`

File: src/main/scala/fr/inria/hopla/ast/ASTFile.scala [hunk 3]
```diff
@@ -38,8 +38,8 @@ abstract case class ASTFile(name: String) {
...
+      astTrait.subFiles = astTrait.subFiles + this
```
Comment (rouvoy) [inline src/main/scala/fr/inria/hopla/ast/ASTFile.scala]: `astTrait.subFiles += this`

**Comment by JonathanGeoffroy** [comment]
The second commit improves source code for the three lines mentioned. Please, feel free to suggest any other improvement before merging. Next step is to write documentation for parser and AST parts, before writing a DSL code generator for our AST.

**Pull request merged**